File: modules/ramfs/src/ramfs.c
#include "../../include/module.h"
#include "./printf.h"
// It may sound funny that both arc_ramfs and arc_inode are weak pointers.
// This is because they are indeed moved out, instead of holding a clone.
// You may see them as &Arc<T> (with proper lifetime automatically), which is easier to understand.
struct ramfs{
    void* arc_ramfs;
    struct fsinfo info;
    struct ramfs_inode* root;
};
struct ramfs_dirent{
    char name[256];
    struct ramfs_inode* item; // Points to arc.
};
struct ramfs_inode{
    struct inode_metadata metadata;
// There are two situations for an inode.
// 1. This is the root inode. In this case the inode works as a "static" one and requires manual recycling when filesystem is released.
// 2. This is an inode attached to some other inode. In this case, it is created with two links (one for returning and one for linking), and works as proper as regular Arcs. So using Arc for linking and unlinking gets the right result.
// In both cases, arc_inode can be seen "attached onto" an existed Arc, so it has same lifecycle as the inode itself(which is trivial, since self-reference always lives as long as the object).
    struct ramfs* parent;    
    void* arc_inode;
    // And overlaying variables
    size_t buf_len;
    char* data;
    size_t links; //This does not matter when recycling.
};
int strcmp(const char* a, const char* b){
    while(1){
        if(*a==*b && *b=='\0') return 0;
        if(*a==*b){
            a++;
            b++;
        }else{
            return 1;
        }
    }
}
size_t strncpy(char* dst, const char* src, size_t n){
    off_t i;
    for(i=0;i<n;i++){
        dst[i]=src[i];
        if(src[i]=='\0') break;
    }
    return i;
}
ssize_t ramfs_resize(void* inode, size_t len){
    char buffer[1024];
    snprintf(buffer, 1024, "ramfs_resize: %ld %ld", inode, len);
    lkm_api_info(buffer);
    struct ramfs_inode* fi=inode;
    if(len==fi->buf_len) return 0;
    if(len==0){
        if(fi->data!=0) lkm_api_kfree(fi->data, fi->buf_len);
        fi->data=0;
        fi->buf_len=0;
        return 0;
    }
    char* new_buffer=lkm_api_kmalloc(len);
    if(new_buffer==0){
        return -1;
    }
    if(fi->buf_len>len){
        //shrinking
        off_t i;
        for(i=0;i<len;i++){
            new_buffer[i]=fi->data[i];
        }
    }else{
        //expanding
        off_t i;
        for(i=0;i<fi->buf_len;i++){
            new_buffer[i]=fi->data[i];        
        }
        for(;i<len;i++){
            new_buffer[i]=0; //Zeroing
        }
    }
    if(fi->data!=0) lkm_api_kfree(fi->data, fi->buf_len);
    fi->data=new_buffer;
    fi->buf_len=len;
    return 0;
}
ssize_t ramfs_find(void* inode, const char* name, size_t len,
 void** /* Arc */ result){
    struct ramfs_inode* fi=inode;
    if(fi->metadata.type_!=1) return -4;
    struct ramfs_dirent* entries=fi->data;
    size_t total_files=fi->buf_len/sizeof(struct ramfs_dirent);
    off_t i;
    for(i=0;i<total_files;i++){
	lkm_api_info(entries[i].name);
        if(strcmp(entries[i].name, name)==0){
            //found.
            *result=lkm_api_clone_arc_inode(entries[i].item->arc_inode);
            return 0;
        }
    }
    return -5;
}
struct ramfs_inode* alloc_inode(struct ramfs* fs){
    struct ramfs_inode* inode=lkm_api_kmalloc(sizeof(struct ramfs_inode));
    off_t i;
    for(i=0;i<sizeof(struct ramfs_inode);i++){
        ((char*)inode)[i]=0;
    }
    lkm_api_info("create_arc_inode");
    inode->arc_inode=lkm_api_create_arc_inode(fs->arc_ramfs, inode);
    inode->parent=fs;
    return inode;
}
void free_inode(struct ramfs_inode* inode){
    ramfs_resize(inode, 0);
    lkm_api_kfree(inode, sizeof(*inode));
}


ssize_t ramfs_read_at(void* inode, size_t offset, void* buf, size_t len){
    struct ramfs_inode* fi=inode;
    size_t remaining=fi->buf_len-offset;
    size_t read_len=(remaining<len?remaining:len);
    off_t off;
    char* cbuf=buf;
    for(off=0;off<read_len;off++){
        cbuf[off]=fi->data[offset+off];
    }
    return read_len;
}
ssize_t ramfs_write_at(void* inode, size_t offset, const void* buf, size_t len){
    char buffer[1024];
    snprintf(buffer, 1024, "ramfs_write_at: %ld %ld %ld %ld", inode, offset, buf, len);
    lkm_api_info(buffer);
    struct ramfs_inode* fi=inode;
    size_t max_size=offset+len;
    if(max_size>fi->buf_len){
        ramfs_resize(inode, max_size);
    }
    lkm_api_info("copy");
    off_t off=0;
    const char* cbuf=buf;
    for(off=0;off<len;off++){
        fi->data[offset+off]=cbuf[off];
    }
    lkm_api_info("copy done");
    return len;
}
ssize_t ramfs_metadata(void* inode, struct inode_metadata* metadata){

    struct ramfs_inode* fi=inode;
    char buffer[1024];
    snprintf(buffer, 1024, "ramfs_metadata: %ld %ld %ld", inode, metadata, fi->metadata.type_);
    lkm_api_info(buffer);
    *metadata=fi->metadata;
    metadata->size=fi->buf_len;
    metadata->blk_size=4096;
    metadata->blocks=((fi->buf_len+4095)/4096);
    metadata->nlinks=fi->links;
    metadata->inode=inode;
    return 0;
}
ssize_t ramfs_set_metadata(void* inode, const struct inode_metadata* metadata){
    struct ramfs_inode* fi=inode;
    fi->metadata=*metadata;
}
ssize_t ramfs_poll(void* inode, struct poll_status* status){
    struct ramfs_inode* fi=inode;
    status->flags=3;
    return 0;
}
ssize_t ramfs_sync_all(void* inode){
    struct ramfs_inode* fi=inode;
    return 0;
}
ssize_t ramfs_sync_data(void* inode){
    struct ramfs_inode* fi=inode;
    return 0;
}


void folder_append_item(struct ramfs_inode* fi, struct ramfs_dirent* dirent){
    ramfs_write_at(fi, fi->buf_len, dirent, sizeof(*dirent));
}
void folder_remove_item(struct ramfs_inode* fi, off_t index){
    struct ramfs_dirent* entries=fi->data;
    size_t total_files=fi->buf_len/sizeof(struct ramfs_dirent);
    entries[index]=entries[total_files-1];
    ramfs_resize(fi, fi->buf_len-sizeof(struct ramfs_dirent));
}
void init_folder(struct ramfs_inode* fi, struct ramfs_inode* parent){
    fi->metadata.type_=1; // Set metadata as directory.
    struct ramfs_dirent init_dirent;
    init_dirent.name[0]='.';
    init_dirent.name[1]='\0';
    init_dirent.item=fi; // This does not count as a reference.
    folder_append_item(fi, &init_dirent);
    init_dirent.name[1]='.';
    init_dirent.name[2]='\0';
    init_dirent.item=parent; // This does not count as a reference, either.
    // A folder can only be removed empty, and can only be linked once.
    // So removing the folder without hesitate about reference counting is safe.
    folder_append_item(fi, &init_dirent);
}

ssize_t ramfs_create(void* inode, const char* name, size_t len, uint64_t type_,
   uint32_t mode, void** result){
    struct ramfs_inode* fi=inode;
    if(fi->metadata.type_!=1) return -4;
    struct ramfs_dirent* entries=fi->data;
    size_t total_files=fi->buf_len/sizeof(struct ramfs_dirent);
    off_t i;
    for(i=0;i<total_files;i++){
        if(strcmp(entries[i].name, name)==0){
            return -6;
        }
    }
    struct ramfs_inode* new_inode=alloc_inode(fi->parent);
    new_inode->metadata.type_=type_;
    if(type_==1){
        init_folder(new_inode, fi);
    }
    struct ramfs_dirent tmpent;
    strncpy(tmpent.name, name, 256);
    tmpent.item=new_inode;
    folder_append_item(fi, &tmpent);
    new_inode->links=1;
    *result=lkm_api_clone_arc_inode(new_inode->arc_inode);
    return 0;
}
ssize_t ramfs_setrdev(void* inode, uint64_t dev){
    struct ramfs_inode* fi=inode;
    fi->metadata.rdev=dev;
    return 0;
}
ssize_t ramfs_unlink(void* inode, const char* name, size_t len){
    if(strcmp(name, ".")==0) return -3;
    if(strcmp(name, "..")==0) return -3;
    struct ramfs_inode* fi=inode;
    if(fi->metadata.type_!=1) return -4;
    struct ramfs_dirent* entries=fi->data;
    size_t total_files=fi->buf_len/sizeof(struct ramfs_dirent);
    off_t i;
    for(i=0;i<total_files;i++){
        if(strcmp(entries[i].name, name)==0){
            //found.
            struct ramfs_inode* target=entries[i].item;
            if(target->metadata.type_==1){
                if(target->buf_len>sizeof(struct ramfs_dirent)*2){
                    return -11; // Directory not empty!
                }
                lkm_api_release_arc_inode(target->arc_inode);
                folder_remove_item(fi, i);
                return 0;
            }else{
                target->links--;
                lkm_api_release_arc_inode(target->arc_inode);
                folder_remove_item(fi, i);
                return 0;
            }
        }
    }
    return -5;
}
ssize_t ramfs_link(void* inode, const char* name, size_t len, void* other){
    struct ramfs_inode* fi=inode;
    struct ramfs_inode* other_fi=other;
    if(other_fi->metadata.type_==1) return -3;
    if(fi->metadata.type_!=1) return -4;
    void* temp;
    if(ramfs_find(inode, name, len, &temp)==0){
        lkm_api_release_arc_inode(temp);
        return -6;
    }
    lkm_api_clone_arc_inode(other_fi->arc_inode); //increase arc.
    // and "bind" it to the folder.
    struct ramfs_dirent tmpent;
    strncpy(tmpent.name, name, 256);
    tmpent.item=other_fi;
    folder_append_item(fi, &tmpent);
    return 0;
}
ssize_t ramfs_move_(void* inode, const char* old_name, size_t old_len,
  void* target, const char* new_name, size_t new_len){
    struct ramfs_inode* fi=inode;
    struct ramfs_inode* other_fi=target;
    if(fi->metadata.type_!=1) return -4;
    if(other_fi->metadata.type_!=1) return -4;
    struct ramfs_dirent* entries=fi->data;
    size_t total_files=fi->buf_len/sizeof(struct ramfs_dirent);
    off_t i;
    for(i=0;i<total_files;i++){
        if(strcmp(entries[i].name, old_name)==0){
            //found.
            void* temp;
            if(ramfs_find(other_fi, new_name, new_len, &temp)==0){
                lkm_api_release_arc_inode(temp); //duplicate and can't move
                return -6;
            }
            struct ramfs_dirent tmpent=entries[i];
            // appending self element again may crash while vector-moving.
            // we copy entries[i] out to prevent this.
            folder_append_item(other_fi, &tmpent);
            folder_remove_item(fi, i);
            return 0;
        }
    }
    return -5;
}

ssize_t ramfs_get_entry(void* inode, size_t id, void* buffer){
    struct ramfs_inode* fi=inode;
    if(fi->metadata.type_!=1) return -4;
    struct ramfs_dirent* entries=fi->data;
    size_t total_files=fi->buf_len/sizeof(struct ramfs_dirent);
    if(id>=total_files){
        return -5;
    }
    strncpy(buffer, &entries[id], 256);
    return 0;
}
ssize_t ramfs_io_control(void* inode, uint32_t cmd, void* data){
    struct ramfs_inode* fi=inode;
    return -1;
}
void ramfs_drop(void* inode){
    struct ramfs_inode* fi=inode; 
    // For ramfs: now should do the cleanup.
    free_inode(inode);
}

struct inode_operations ramfs_inode_ops={
    .read_at=ramfs_read_at,
    .write_at=ramfs_write_at,
    .metadata=ramfs_metadata,
    .poll=ramfs_poll,
    .sync_all=ramfs_sync_all,
    .sync_data=ramfs_sync_data,
    .resize=ramfs_resize,
    .create=ramfs_create,
    .setrdev=ramfs_setrdev,
    .unlink=ramfs_unlink,
    .link=ramfs_link,
    .move_=ramfs_move_,
    .find=ramfs_find,
    .get_entry=ramfs_get_entry,
    .io_control=ramfs_io_control,
    .drop=ramfs_drop
};

ssize_t ramfs_fs_mount(uint64_t flags, const char* dev_name, void* data,
                     void* arc_fs, void** result){
    char buffer[1024];
    snprintf(buffer, 1024, "mounting onto %ld\n", arc_fs);
    lkm_api_info("Start mounting");
    lkm_api_info(buffer);
    struct ramfs* fs=lkm_api_kmalloc(sizeof(struct ramfs*));
    lkm_api_info("malloc");
    off_t i;
    lkm_api_info("memset");
    for(i=0;i<sizeof(*fs);i++) ((char*)fs)[i]=0; //memset
    fs->arc_ramfs=arc_fs;
    lkm_api_info("alloc_inode");
    struct ramfs_inode* root=alloc_inode(fs);
    root->links=1;
    lkm_api_info("init_folder");
    init_folder(root, root);
    lkm_api_info("set");
    fs->root=root;
    *result=fs;
    lkm_api_info("done");
    return 0;
}
ssize_t ramfs_fs_sync(void* fs){
    return 0;
}
ssize_t ramfs_fs_root_inode(void* fs, void** /* Arc */ inode){
    struct ramfs* rfs=fs;
    *inode=lkm_api_clone_arc_inode(rfs->root->arc_inode);
    return 0;
}
ssize_t ramfs_fs_info(void* fs, struct fsinfo* info){
    struct ramfs* rfs=fs;
    return 0;
}
void ramfs_fs_drop(void* fs){
    lkm_api_info("Drop FS!");
    struct ramfs* rfs=fs;
    lkm_api_release_arc_inode(rfs->root->arc_inode);
    lkm_api_kfree(fs, sizeof(struct ramfs));
}
struct filesystem_operations ramfs_filesystem_ops={
    .mount=ramfs_fs_mount,
    .sync=ramfs_fs_sync,
    .root_inode=ramfs_fs_root_inode,
    .info=ramfs_fs_info,
    .drop=ramfs_fs_drop
};
void _putchar(char chr){/*unused*/}
void init_module(){
    char buffer[1024];
    snprintf(buffer, 1024, "external function: %ld %ld", ramfs_inode_ops, ramfs_inode_ops.metadata);
    lkm_api_info(buffer);
    lkm_api_register_fs("ramfs", &ramfs_filesystem_ops, &ramfs_inode_ops, 0);   
}

File: modules/ramfs/src/main/ramfs.rs
extern crate rcore;
extern crate alloc;
extern crate spin;
extern crate rcore_fs;
use rcore::lkm::fs::FileSystemType;
use alloc::sync::{Arc, Weak};
use rcore_fs::vfs::{FileSystem, FsError, INode, PollStatus, Metadata, FileType, FsInfo, Timespec};
use alloc::string::{String, ToString};
use core::any::Any;
use alloc::vec::Vec;
use spin::{RwLock, RwLockWriteGuard};
use rcore::lkm::fs::FileSystemManager;
use core::cmp::min;
use alloc::vec;
use rcore::syscall::Syscall;
pub struct RamFSBehav{}
impl RamFSBehav{
    pub fn registerRamFS(){
        FileSystemManager::get().write().registerFileSystem("ramfs", RamFSBehav{});
    }
}
struct LockedINode(RwLock<RamFSINode>);
impl FileSystemType for RamFSBehav{
    fn mount(&self, syscall: &mut Syscall, source: &str, flags: u64, data: usize) -> Result<Arc<FileSystem>, FsError> {
        let root=Arc::new(LockedINode(RwLock::new(RamFSINode{
            this: None,
            parent: None,

            children: Vec::new(),
            content: Vec::new(),
            extra: Metadata{
                dev: 0,
                inode: 0,
                size: 0,
                blk_size: 0,
                blocks: 0,
                atime: Timespec { sec: 0, nsec: 0 },
                mtime: Timespec { sec: 0, nsec: 0 },
                ctime: Timespec { sec: 0, nsec: 0 },
                type_: FileType::Dir,
                mode: 0,
                nlinks: 1,
                uid: 0,
                gid: 0,
                rdev: 0
            },
            fs: Weak::default()
        })));
        let mut fs=Arc::new(RamFS{
            root: root
        });
        let mut root=downcast_inode(Arc::as_ref(&fs.root)).0.write();
        root.parent=Some(Arc::downgrade(&fs.root));
        root.this=Some(Arc::downgrade(&fs.root));
        root.fs=Arc::downgrade(&fs);
        root.extra.inode=Arc::into_raw(root.this.as_ref().unwrap().upgrade().unwrap()) as *const RamFSINode as usize;
        drop(root);
        Ok(fs)

    }
}

pub struct RamFS{
    root: Arc<INode>
}

impl FileSystem for RamFS{
    fn sync(&self) -> Result<(), FsError> {
        Ok(())
    }

    fn root_inode(&self) -> Arc<INode> {
        Arc::clone(&self.root)
    }

    fn info(&self) -> FsInfo {
        FsInfo{
            bsize: 0,
            frsize: 0,
            blocks: 0,
            bfree: 0,
            bavail: 0,
            files: 0,
            ffree: 0,
            namemax: 0
        }
    }
}
pub struct RamFSINode{
    parent: Option<Weak<INode>>,
    this: Option<Weak<INode>>,
    children: Vec<(String, Arc<INode>)>,
    content: Vec<u8>,
    extra: Metadata,
    fs: Weak<RamFS>
}

fn lockMultiple<'a>(locks: &[&'a RwLock<RamFSINode>])->Vec<RwLockWriteGuard<'a, RamFSINode>>{
    let mut v: Vec<(usize, &'a RwLock<RamFSINode>, usize)>=Vec::new();
    let mut index:usize=0;
    for l in locks{
        v.push((index, *l, l.read().extra.inode));
        index+=1;
    }
    v.sort_by_key(|l| l.2);
    let mut v2: Vec<(usize, RwLockWriteGuard<'a, RamFSINode>)>=v.into_iter().map(|(index, lock, inode)| (index, lock.write())).collect();
    v2.sort_by_key(|(index, lock)| *index);
    v2.into_iter().map(|(index, lock)| lock).collect()
}
fn downcast_inode<'a>(inode: &'a INode)->&'a LockedINode{
    inode.as_any_ref().downcast_ref::<LockedINode>().unwrap()
}
impl INode for LockedINode {
    fn read_at(&self, offset: usize, buf: &mut [u8]) -> Result<usize, FsError> {
        let file = self.0.read();
        if file.extra.type_ == FileType::Dir {
            return Err(FsError::IsDir);
        }
        let src = &file.content[min(offset, file.content.len())..min(offset + buf.len(), file.content.len())];
        let target = &buf[0..src.len()];
        buf.copy_from_slice(src);
        Ok(src.len())
    }

    fn write_at(&self, offset: usize, buf: &[u8]) -> Result<usize, FsError> {
        let mut file = self.0.write();
        if file.extra.type_ == FileType::Dir {
            return Err(FsError::IsDir);
        }
        let content = &mut file.content;
        if offset + buf.len() > content.len() {
            content.resize(offset + buf.len(), 0);
        }
        let target = &mut content[offset..offset + buf.len()];
        target.copy_from_slice(buf);
        Ok(buf.len())
    }

    fn poll(&self) -> Result<PollStatus, FsError> {
        let file = self.0.read();
        if file.extra.type_ == FileType::Dir {
            return Err(FsError::IsDir);
        }
        Ok(PollStatus {
            read: true,
            write: true,
            error: false
        })
    }

    fn metadata(&self) -> Result<Metadata, FsError> {
        let file = self.0.read();
        let extra = &file.extra;
        let size = file.content.len();
        Ok(Metadata {
            dev: 0,
            inode: extra.inode,
            size: size,
            blk_size: 4096,
            blocks: size / 4096,
            atime: extra.atime,
            mtime: extra.mtime,
            ctime: extra.ctime,
            type_: extra.type_,
            mode: extra.mode,
            nlinks: extra.nlinks,
            uid: extra.uid,
            gid: extra.gid,
            rdev: extra.rdev
        })
    }

    fn set_metadata(&self, metadata: &Metadata) -> Result<(), FsError> {
        let mut file = self.0.write();
        file.extra.atime = metadata.atime;
        file.extra.mtime = metadata.mtime;
        file.extra.ctime = metadata.ctime;
        file.extra.mode = metadata.mode;
        file.extra.uid = metadata.uid;
        file.extra.gid = metadata.gid;
        Ok(())
    }

    fn sync_all(&self) -> Result<(), FsError> {
        Ok(())
    }

    fn sync_data(&self) -> Result<(), FsError> {
        Ok(())
    }

    fn resize(&self, len: usize) -> Result<(), FsError> {
        let mut file = self.0.write();
        if file.extra.type_ == FileType::File {
            file.content.resize(len, 0);
            Ok(())
        } else {
            Err(FsError::NotFile)
        }
    }

    fn create(&self, name: &str, type_: FileType, mode: u32) -> Result<Arc<INode>, FsError> {
        let mut file = self.0.write();
        if file.extra.type_ == FileType::Dir {
            if name == "." { return Err(FsError::EntryExist); }
            if name == ".." { return Err(FsError::EntryExist); }
            for (n, _) in file.children.iter() {
                if n == name {
                    return Err(FsError::EntryExist);
                }
            }
            let temp_file: Arc<INode> = Arc::new(LockedINode(RwLock::new(RamFSINode {
                parent: Some(Weak::clone(file.this.as_ref().unwrap())),
                this: None,
                children: Vec::new(),
                content: Vec::new(),
                extra: Metadata {
                    dev: 0,
                    inode: 0,
                    size: 0,
                    blk_size: 0,
                    blocks: 0,
                    atime: Timespec { sec: 0, nsec: 0 },
                    mtime: Timespec { sec: 0, nsec: 0 },
                    ctime: Timespec { sec: 0, nsec: 0 },
                    type_: type_,
                    mode: mode as u16,
                    nlinks: 0,
                    uid: 0,
                    gid: 0,
                    rdev: 0
                },
                fs: Weak::clone(&file.fs)
            })));
            let mut root = downcast_inode(Arc::as_ref(&temp_file)).0.write();
            root.this = Some(Arc::downgrade(&temp_file));
            drop(root);
            file.children.push((String::from(name), Arc::clone(&temp_file)));
            Ok(temp_file)
        } else {
            Err(FsError::NotDir)
        }
    }

    fn link(&self, name: &str, other: &Arc<INode>) -> Result<(), FsError> {
        let other_lock = downcast_inode(Arc::as_ref(other));
        // to make sure locking order.
        let mut locks = lockMultiple(&vec![&self.0, &other_lock.0]).into_iter();

        let mut file = locks.next().unwrap();
        let mut other_l = locks.next().unwrap();

        if file.extra.type_ != FileType::Dir {
            return Err(FsError::NotDir);
        }
        if other_l.extra.type_ == FileType::Dir {
            return Err(FsError::IsDir)
        }
        for (n, _) in file.children.iter() {
            if n == name {
                return Err(FsError::EntryExist);
            }
        }

        file.children.push((String::from(name), Arc::clone(other)));
        other_l.extra.nlinks += 1;
        Ok(())
    }

    fn unlink(&self, name: &str) -> Result<(), FsError> {
        let mut file = self.0.write();
        if file.extra.type_ != FileType::Dir {
            return Err(FsError::NotDir);
        }
        if name == "." || name == ".." {
            return Err(FsError::DirNotEmpty);
        }
        let mut index: usize = 0;
        for (n, f) in file.children.iter() {
            if n == name {
                let removal_inode = Arc::clone(f);
                let other = downcast_inode(Arc::as_ref(&removal_inode));
                if other.0.read().children.len() > 0 {
                    return Err(FsError::DirNotEmpty);
                }
                file.children.remove(index);
                drop(file);
                other.0.write().extra.nlinks -= 1;
                return Ok(())
            } else {
                index += 1;
            }
        }
        Ok(())
    }

    fn move_(&self, old_name: &str, target: &Arc<INode>, new_name: &str) -> Result<(), FsError> {
        let elem = self.find(old_name)?;
        let t = downcast_inode(Arc::as_ref(target));
        if let Err(err) = t.link(new_name, &elem) {
            return Err(err);
        } else {
            if let Err(err) = self.unlink(old_name) {
                t.unlink(new_name)?;
                return Err(err);
            } else {
                return Ok(());
            }
        }
    }

    fn find(&self, name: &str) -> Result<Arc<INode>, FsError> {
        let file = self.0.read();
        if file.extra.type_ != FileType::Dir {
            return Err(FsError::NotDir);
        }
        //info!("find it: {} {}", name, file.parent.is_none());
        match name {
            "." => Ok(file.this.as_ref().unwrap().upgrade().ok_or(FsError::EntryNotFound)?),
            ".." => Ok(file.parent.as_ref().unwrap().upgrade().ok_or(FsError::EntryNotFound)?),
            name => {
                for (s, e) in file.children.iter() {
                    if s == name {
                        return Ok(Arc::clone(e));
                    }
                }
                Err(FsError::EntryNotFound)
            }
        }
    }

    fn get_entry(&self, id: usize) -> Result<String, FsError> {
        let file = self.0.read();
        if file.extra.type_ != FileType::Dir {
            return Err(FsError::NotDir);
        }

        match id {
            0 => Ok(String::from(".")),
            1 => Ok(String::from("..")),
            i => {
                if let Some((s, _)) = file.children.get(i - 2) {
                    Ok(s.to_string())
                } else {
                    Err(FsError::EntryNotFound)
                }
            }
        }
    }

    fn io_control(&self, cmd: u32, data: usize) -> Result<(), FsError> {
        Err(FsError::NotSupported)
    }

    fn fs(&self) -> Arc<FileSystem> {
        Weak::upgrade(&self.0.read().fs).unwrap()
    }

    fn as_any_ref(&self) -> &Any {
        self
    }
/*
    fn setrdev(&self, dev: u64) -> Result<(), FsError> {
        self.0.write().extra.rdev = dev;
        Ok(())
    }
*/
}
